smt.go

```diff
@@ -0,0 +1,207 @@
+	emptyLeafHash                  []byte
+	levelsUpHashOfEmptyLeaves      []Hash
+	treeHeight                     uint
+	noOfNonEmptyLeaves             int
+}
+
+func NewSMTWithTwoHashFuncs(leafHashFunc hash.Hash, nonLeafHashFunc hash.Hash) (SMT, error) {
+	emptyLeafHash, err := emptyLeafHash(leafHashFunc)
+	if err != nil {
+		return SMT{}, err
+	}
+	return newSMT(leafHashFunc, nonLeafHashFunc, emptyLeafHash)
+}
+
+func NewSMTWithNonLeafHashAndEmptyLeafHashValue(emptyLeafHash []byte, nonLeafHashFunc hash.Hash) (SMT, error) {
+	return newSMT(nil, nonLeafHashFunc, emptyLeafHash)
+}
+
+func (self *SMT) RootHash() []byte {
+	if self.noOfNonEmptyLeaves == 0 {
+		return self.levelsUpHashOfEmptyLeaves[len(self.levelsUpHashOfEmptyLeaves)-1]
+	}
+	return self.nodesWithoutEmptyLeavesSubTree[self.treeHeight-1][0]
+}
+
+func (self *SMT) Generate(nonEmptyLeaves [][]byte, totalSize uint64) error {
+	if !isPowerOfTwo(totalSize) {
+		return errors.New("Leaves number of SMT tree should be power of 2")
+	}
+	if uint64(len(nonEmptyLeaves)) > totalSize {
+		return errors.New("NonEmptyLeaves is bigger than totalSize ")
+	}
+	self.treeHeight = uint(logBaseTwo(totalSize) + 1)
+	self.noOfNonEmptyLeaves = len(nonEmptyLeaves)
+
+	noOfEmtpyLeaves := totalSize - uint64(len(nonEmptyLeaves))
+	maxEmtySubTreeHeight := 0
+	for i := noOfEmtpyLeaves; i > 0; i = i >> 1 {
+		maxEmtySubTreeHeight++
+	}
+	err := self.computeEmptyLeavesSubTreeHash(maxEmtySubTreeHeight)
+	if err != nil {
+		return err
+	}
+	return self.buildAllLevelNodes(nonEmptyLeaves)
+}
+
+func (self *SMT) GetMerkelProof(leafNo uint) []ProofNode {
+	proofs := []ProofNode{}
+	level := int(self.treeHeight - 1)
+	index := leafNo
+	for i := level; i > 0; i-- {
+		proofNode := self.proofNodeAt(int(index), int(i))
+		proofs = append(proofs, proofNode)
+		index = index / 2
+	}
+	return proofs
+}
+
+//Following are non public function
+func newSMT(leafHashFunc hash.Hash, nonLeafHashFunc hash.Hash, emptyLeafHash []byte) (SMT, error) {
+	smt := SMT{nodesWithoutEmptyLeavesSubTree: [][]Hash{}, levelsUpHashOfEmptyLeaves: []Hash{emptyLeafHash}, emptyLeafHash: emptyLeafHash, leafHashFunc: leafHashFunc, nonLeafHashFunc: nonLeafHashFunc}
+	return smt, nil
+}
+
+func (self *SMT) computeEmptyLeavesSubTreeHash(maxHeight int) error {
+	lastLevelHash := self.emptyLeafHash
+	var err error
+	for i := 1; i < maxHeight; i++ {
+		lastLevelHash, err = self.parentHash(lastLevelHash, lastLevelHash)
+		if err != nil {
+			return err
+		}
+		self.levelsUpHashOfEmptyLeaves = append(self.levelsUpHashOfEmptyLeaves, lastLevelHash)
+	}
+	return nil
+}
+
+func (self *SMT) buildAllLevelNodes(nonEmptyLeaves [][]byte) error {
+	err := self.buildLeavesNodes(nonEmptyLeaves)
+	if err != nil {
+		return err
+	}
+	for i := self.treeHeight; i > 1; i-- {
+		err := self.buildInternalOneLevelNodes(i - 1)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (self *SMT) buildLeavesNodes(nonEmptyLeaves [][]byte) error {
```

Comment:
I think we shouldn't hash the leaves in the SMT itself, that's a lot less flexible than hashing the leaves before. We should hash the leaves in precise-proofs instead.